From 0af9af8890bc5f818a18d750d8915b7ff90c79c2 Mon Sep 17 00:00:00 2001
From: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
Date: Sat, 27 Jun 2015 20:55:47 +0300
Subject: [PATCH 03/13] serial: sh-sci: Fix race condition of SCSCR register
 access

This fixes race condition in concurrent recieve/transmit using DMA.

Signed-off-by: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
Signed-off-by: Kazuya Mizuguchi <kazuya.mizuguchi.ks@renesas.com>
---
 drivers/tty/serial/sh-sci.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/tty/serial/sh-sci.c b/drivers/tty/serial/sh-sci.c
index e573282..abfbce3 100644
--- a/drivers/tty/serial/sh-sci.c
+++ b/drivers/tty/serial/sh-sci.c
@@ -1719,7 +1719,12 @@ static void rx_timer_fn(unsigned long arg)
 {
 	struct sci_port *s = (struct sci_port *)arg;
 	struct uart_port *port = &s->port;
-	u16 scr = serial_port_in(port, SCSCR);
+	u16 scr;
+	unsigned long flags;
+
+	spin_lock_irqsave(&port->lock, flags);
+
+	scr = serial_port_in(port, SCSCR);
 
 	if (port->type == PORT_SCIFA || port->type == PORT_SCIFB) {
 		scr &= ~SCSCR_RDRQE;
@@ -1727,6 +1732,9 @@ static void rx_timer_fn(unsigned long arg)
 	}
 	serial_port_out(port, SCSCR, scr | SCSCR_RIE);
 	dev_dbg(port->dev, "DMA Rx timed out\n");
+
+	spin_unlock_irqrestore(&port->lock, flags);
+
 	schedule_work(&s->work_rx);
 }
 
-- 
1.9.1

